File: lib/rice_cooker.dart
import 'dart:io';

class RiceCooker {
  late bool power;

  RiceCooker() {
    power = false;
  }

  bool checkPower() {
    return power;
  }

  void turnOn() {
    power = true;
    print("The rice cooker is turned on.");
  }

  void turnOff() {
    power = false;
    print("The rice cooker is turned off.");
  }

  void chooseFunction(int action, int? duration, int? temperature) {
    if (action == 1) {
      turnOn();
      try {
        switch (action) {
          case 1:
            if (duration == null || temperature == null) {
              throw ArgumentError("Duration and temperature must be provided.");
            }
            cook(duration, temperature);
            break;
          case 2:
            boil();
            break;
          default:
            print('Unrecognized action.');
        }
      } catch (error) {
        print('Error: $error');
      }
    } else if (action == 2) {
      turnOff();
    } else {
      print('Unrecognized command.');
    }
  }

  void cook(int duration, int temperature) {
    try {
      print('Cooking in progress for $duration minute(s) at $temperature°C...');
      sleep(Duration(minutes: duration));
      print('Cooking completed!');
    } catch (error) {
      print('Error: $error');
    }
  }

  void boil() {
    try {
      print('Boiling water in progress...');
      const boilingTime = 10; // Simulation duration (10 minutes, for example)
      int currentTemperature = 0;

      for (int i = 1; i <= boilingTime; i++) {
        currentTemperature += 10;
        sleep(Duration(minutes: 1)); // Simulate the time needed to increase the temperature
        print('Water temperature: $currentTemperature°C - Remaining time: ${boilingTime - i} minutes');
      }

      print('Water has reached 100°C. Automatic shutdown.');
    } catch (error) {
      print('Error: $error');
    }
  }
}

void main() {
  RiceCooker myRiceCooker = RiceCooker();

  stdout.write('Hello! You are using the rice cooker. Do you want to turn it on? (1 for yes / 2 for no): ');
  int userDecision = int.parse(stdin.readLineSync()!);

  if (userDecision == 1) {
    stdout.write('What do you want to do? (1 to cook / 2 to boil): ');
    int action = int.parse(stdin.readLineSync()!);

    if (action == 1) {
      stdout.write('How many minutes do you want to cook? ');
      int duration = int.parse(stdin.readLineSync()!);

      stdout.write('At what temperature do you want to cook? ');
      int temperature = int.parse(stdin.readLineSync()!);

      myRiceCooker.chooseFunction(action, duration, temperature);
    } else {
      myRiceCooker.chooseFunction(action, null, null);
    }
  } else {
    myRiceCooker.chooseFunction(2, null, null);
  }
}

File: src/RiceCooker.cpp
#include "RiceCooker.h"
#include <iostream>

RiceCooker::RiceCooker() : power(false) {}

bool RiceCooker::checkPower() const {
    return power;
}

void RiceCooker::turnOn() {
    power = true;
    printMessage("The rice cooker is turned on.");
}

void RiceCooker::turnOff() {
    power = false;
    printMessage("The rice cooker is turned off.");
}

void RiceCooker::chooseFunction(int choice) {
    if (!power) {
        printMessage("The rice cooker is not turned on.");
        return;
    }

    int userChoice;
    while (true) {
        printMessage("Choose a function (1 to cook rice, 2 to heat water, 3 to heat food, 4 to turn off): ");

        if (!(std::cin >> userChoice)) {
            printMessage("Error: Please enter an integer.");
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            continue;
        }

        if (userChoice < 1 || userChoice > 4) {
            printMessage("Error: Invalid choice. Please choose a number between 1 and 4.");
            continue;
        }

        break;
    }

    switch (userChoice) {
        case 1:
            int time;
            while (true) {
                printMessage("Enter the cooking duration in minutes: ");
                if (!(std::cin >> time)) {
                    printMessage("Error: Please enter an integer.");
                    std::cin.clear();
                    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
                    continue;
                }

                break;
            }
            cookRice(time);
            break;
        case 2:
            heatWater();
            break;
        case 3:
            heatFood();
            break;
        case 4:
            turnOff();
            break;
        default:
            printMessage("Unrecognized choice.");
    }
}

void RiceCooker::cookRice(int time) {
    printMessage("Cooking rice in progress...");
    sleep(time);
    printMessage("The rice is ready!");
}

void RiceCooker::heatWater() {
    printMessage("Heating water in progress...");
    int temperature = 0;
    while (temperature < 100) {
        temperature += 10;
        sleep(1);
        printMessage("Water temperature: " + std::to_string(temperature) + "°C");
    }
    printMessage("Water has reached 100°C. Automatically turning off.");
}

void RiceCooker::heatFood() {
    printMessage("Heating food in progress...");
    printMessage("The food is heated and ready to be consumed!");
}

void RiceCooker::sleep(int minutes) {
    std::chrono::milliseconds milliseconds(minutes * 60 * 1000);
    auto startTime = std::chrono::system_clock::now();
    while (std::chrono::system_clock::now() < startTime + milliseconds);
}

void RiceCooker::printMessage(const std::string& message) const {
    std::cout << message << std::endl;
}

int RiceCooker::getUserInput() const {
    int userInput;
    std::cin >> userInput;
    return userInput;
}
